File: backend/app/engines/debt_strategies.py
"""Debt payoff strategy calculators — port of src/lib/debt-strategies.ts.

Avalanche: highest interest first.
Snowball: smallest balance first.

`date` strings on each step are produced for presentational parity with
the TS engine but are runtime-dependent. Parity tests strip them before
comparing — `month` is the canonical time index.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Callable, Literal

from app.engines._helpers import js_round

_MONTHS_SHORT = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
_MAX_MONTHS = 360  # 30-year safety cap


def _month_to_date(offset: int) -> str:
    now = datetime.now()
    year = now.year
    month_index = now.month - 1 + offset
    year += month_index // 12
    month_index = month_index % 12
    return f"{_MONTHS_SHORT[month_index]} {year}"


def _monthly_interest(balance: float, annual_rate: float) -> float:
    return balance * (annual_rate / 100 / 12)


def _simulate(
    debts: list[dict[str, Any]],
    extra_payment: float,
    sort_fn: Callable[[dict[str, Any]], Any],
) -> dict[str, Any]:
    if not debts:
        return {
            "strategy": "avalanche",
            "totalMonths": 0,
            "totalInterestPaid": 0,
            "totalPaid": 0,
            "steps": [],
            "payoffDates": {},
        }

    active = [
        {
            "id": d["id"],
            "name": d["name"],
            "balance": d["principal"],
            "interestRate": d["interestRate"],
            "minPayment": d["monthlyPayment"],
        }
        for d in debts
    ]

    steps: list[dict[str, Any]] = []
    payoff_dates: dict[str, int] = {}
    total_interest_paid = 0.0
    total_paid = 0.0
    month = 0

    while active and month < _MAX_MONTHS:
        month += 1
        active.sort(key=sort_fn)

        total_min = sum(d["minPayment"] for d in active)
        available = total_min + extra_payment

        for debt in active:
            interest = _monthly_interest(debt["balance"], debt["interestRate"])
            debt["balance"] += interest
            total_interest_paid += interest

            payment = min(debt["balance"], debt["minPayment"])
            debt["balance"] -= payment
            available -= payment
            total_paid += payment

            steps.append({
                "month": month,
                "date": _month_to_date(month),
                "debtId": debt["id"],
                "debtName": debt["name"],
                "payment": payment,
                "remainingBalance": max(0, debt["balance"]),
                "interestPaid": interest,
                "isPaidOff": debt["balance"] <= 0.01,
            })

        # Apply extra payment waterfall to the top-priority debt
        if available > 0 and active:
            target = active[0]
            extra_pay = min(target["balance"], available)
            target["balance"] -= extra_pay
            total_paid += extra_pay

            # Update the last step for this debt
            for step in reversed(steps):
                if step["month"] == month and step["debtId"] == target["id"]:
                    step["payment"] += extra_pay
                    step["remainingBalance"] = max(0, target["balance"])
                    step["isPaidOff"] = target["balance"] <= 0.01
                    break

        # Remove paid-off debts
        new_active = []
        for d in active:
            if d["balance"] <= 0.01:
                payoff_dates[d["id"]] = month
            else:
                new_active.append(d)
        active = new_active

    return {
        "strategy": "avalanche",
        "totalMonths": month,
        "totalInterestPaid": js_round(total_interest_paid),
        "totalPaid": js_round(total_paid),
        "steps": steps,
        "payoffDates": payoff_dates,
    }


def avalanche(debts: list[dict[str, Any]], extra_monthly_payment: float = 0) -> dict[str, Any]:
    result = _simulate(debts, extra_monthly_payment, sort_fn=lambda d: -d["interestRate"])
    result["strategy"] = "avalanche"
    return result


def snowball(debts: list[dict[str, Any]], extra_monthly_payment: float = 0) -> dict[str, Any]:
    result = _simulate(debts, extra_monthly_payment, sort_fn=lambda d: d["balance"])
    result["strategy"] = "snowball"
    return result


def compare_strategies(debts: list[dict[str, Any]], extra_monthly_payment: float = 0) -> dict[str, Any]:
    a = avalanche(debts, extra_monthly_payment)
    s = snowball(debts, extra_monthly_payment)
    recommendation: Literal["avalanche", "snowball"] = (
        "avalanche" if a["totalInterestPaid"] < s["totalInterestPaid"] else "snowball"
    )
    return {
        "avalanche": a,
        "snowball": s,
        "interestSaved": s["totalInterestPaid"] - a["totalInterestPaid"],
        "monthsDifference": s["totalMonths"] - a["totalMonths"],
        "recommendation": recommendation,
    }

File: frontend/src/lib/debt-strategies.ts
// ============================================================
// FinPath — Debt Payoff Strategy Calculators
// Avalanche (highest interest first) & Snowball (lowest balance first)
// ============================================================

import type { DebtItem } from './types';

export interface PayoffStep {
  month: number;
  date: string;
  debtId: string;
  debtName: string;
  payment: number;
  remainingBalance: number;
  interestPaid: number;
  isPaidOff: boolean;
}

export interface PayoffResult {
  strategy: 'avalanche' | 'snowball';
  totalMonths: number;
  totalInterestPaid: number;
  totalPaid: number;
  steps: PayoffStep[];
  /** Month each debt is paid off */
  payoffDates: Record<string, number>;
}

interface DebtState {
  id: string;
  name: string;
  balance: number;
  interestRate: number;
  minPayment: number;
}

/**
 * Calculate monthly interest for a debt
 */
function monthlyInterest(balance: number, annualRate: number): number {
  return balance * (annualRate / 100 / 12);
}

/**
 * Format month number to date string
 */
function monthToDate(monthOffset: number): string {
  const now = new Date();
  const d = new Date(now.getFullYear(), now.getMonth() + monthOffset, 1);
  return d.toLocaleDateString('en-IN', { month: 'short', year: 'numeric' });
}

/**
 * Run a debt payoff simulation with a given ordering strategy
 */
function simulate(
  debts: DebtItem[],
  extraPayment: number,
  sortFn: (a: DebtState, b: DebtState) => number
): PayoffResult {
  if (debts.length === 0) {
    return {
      strategy: 'avalanche',
      totalMonths: 0,
      totalInterestPaid: 0,
      totalPaid: 0,
      steps: [],
      payoffDates: {},
    };
  }

  // Initialize state
  let activeDebts: DebtState[] = debts.map(d => ({
    id: d.id,
    name: d.name,
    balance: d.principal,
    interestRate: d.interestRate,
    minPayment: d.monthlyPayment,
  }));

  const steps: PayoffStep[] = [];
  const payoffDates: Record<string, number> = {};
  let totalInterestPaid = 0;
  let totalPaid = 0;
  let month = 0;
  const maxMonths = 360; // 30 year safety cap

  while (activeDebts.length > 0 && month < maxMonths) {
    month++;
    
    // Sort debts by strategy
    activeDebts.sort(sortFn);

    // Calculate total minimum payments
    const totalMinPayments = activeDebts.reduce((sum, d) => sum + d.minPayment, 0);
    let available = totalMinPayments + extraPayment;

    for (const debt of activeDebts) {
      // Apply interest
      const interest = monthlyInterest(debt.balance, debt.interestRate);
      debt.balance += interest;
      totalInterestPaid += interest;

      // Pay minimum or remaining balance
      const payment = Math.min(debt.balance, debt.minPayment);
      debt.balance -= payment;
      available -= payment;
      totalPaid += payment;

      steps.push({
        month,
        date: monthToDate(month),
        debtId: debt.id,
        debtName: debt.name,
        payment,
        remainingBalance: Math.max(0, debt.balance),
        interestPaid: interest,
        isPaidOff: debt.balance <= 0.01,
      });
    }

    // Apply extra payment to the priority debt (first in sorted order)
    if (available > 0 && activeDebts.length > 0) {
      const target = activeDebts[0];
      const extraPay = Math.min(target.balance, available);
      target.balance -= extraPay;
      totalPaid += extraPay;

      // Update the last step for this debt
      const lastStep = steps.filter(s => s.month === month && s.debtId === target.id).pop();
      if (lastStep) {
        lastStep.payment += extraPay;
        lastStep.remainingBalance = Math.max(0, target.balance);
        lastStep.isPaidOff = target.balance <= 0.01;
      }
    }

    // Remove paid-off debts
    activeDebts = activeDebts.filter(d => {
      if (d.balance <= 0.01) {
        payoffDates[d.id] = month;
        return false;
      }
      return true;
    });
  }

  return {
    strategy: 'avalanche',
    totalMonths: month,
    totalInterestPaid: Math.round(totalInterestPaid),
    totalPaid: Math.round(totalPaid),
    steps,
    payoffDates,
  };
}

/**
 * Avalanche strategy: Pay highest interest rate debts first
 * Minimizes total interest paid
 */
export function avalanche(debts: DebtItem[], extraMonthlyPayment: number = 0): PayoffResult {
  const result = simulate(
    debts,
    extraMonthlyPayment,
    (a, b) => b.interestRate - a.interestRate // Highest interest first
  );
  result.strategy = 'avalanche';
  return result;
}

/**
 * Snowball strategy: Pay smallest balance debts first
 * Psychologically motivating (quick wins)
 */
export function snowball(debts: DebtItem[], extraMonthlyPayment: number = 0): PayoffResult {
  const result = simulate(
    debts,
    extraMonthlyPayment,
    (a, b) => a.balance - b.balance // Smallest balance first
  );
  result.strategy = 'snowball';
  return result;
}

/**
 * Compare both strategies and return both results
 */
export function compareStrategies(debts: DebtItem[], extraMonthlyPayment: number = 0) {
  const avalancheResult = avalanche(debts, extraMonthlyPayment);
  const snowballResult = snowball(debts, extraMonthlyPayment);

  return {
    avalanche: avalancheResult,
    snowball: snowballResult,
    interestSaved: snowballResult.totalInterestPaid - avalancheResult.totalInterestPaid,
    monthsDifference: snowballResult.totalMonths - avalancheResult.totalMonths,
    recommendation: avalancheResult.totalInterestPaid < snowballResult.totalInterestPaid
      ? 'avalanche' as const
      : 'snowball' as const,
  };
}
